### PR TITLE
Fix for writing the same allow list twice during agent activation

### PR DESCRIPTION
**Pull request by galmasi**
Addresses Issue #1146 -- failing tenant add with IMA in unusual circumstances (i.e. unreliable mysql connections, or when attempting to run the same `tenant -c add` command simultaneously from two different tenants).

Signed-off-by: George Almasi <gheorghe@us.ibm.com>

### CODE REVIEW

**Comment by THS-on** [comment]
For the `too-many-nested-statements` you can just add an ignore comment. The other changes LGTM, but some one else should also look at this before merging.

**Comment by aplanas** [comment]
I agree with @THS-on . Add the ignore and a TODO to mark it. The fix in the first commit is direct, and the refactoring because of the complain is hiding it.

You can later split the rest in a different commit and remove the ignore. I will review also the second one (but I think that I get how you addressed it)

**Comment by galmasi** [comment]
I was looking at the "nested" lint issue and found that the error condition handling in this function is all over the place -- some if statements defer failures to the end of the function, others return immediately.

I saw an opportunity for fixing the indentation issue by front-loading all failures and leaving the rest of the code unindented. But this results in 100+ lines changed for what was originally a 2 line PR.

In the end, it's a choice between a minimalistic *semantic* fix and a fairly large *stylistic* fix. You get to choose. If you don't like the latest evolution of the PR I start from scratch with the 2 lines that fix the actual problem.

**Comment by galmasi** [comment]
FWIW even the 2 line semantic fix leaves out the fact that some variables are potentially undefined, depending on the path taken through the code. I would urge a serious review of this function (`cloud_verifier_tornado::post`) because there be dragons there. But I didn't want to broach the larger subject in this PR.

**Comment by galmasi** [comment]
Al right @aplanas @THS-on I hear you. Going back to the 2-line fix with a lint exception soon(ish). Don't merge anything yet :)

**Comment by THS-on** [comment]
@galmasi I agree that we should refactor this method to make it cleaner. So it  still makes sense to keep your commits and just put them into another PR. Ideally we get https://github.com/keylime/keylime/pull/1139 merged before which also addresses some of the issues.

**Review by mbestavros — CHANGES_REQUESTED** [review]
We already perform a check for allowlists that exist in the DB to prevent overwriting, but you're right that it doesn't catch the case of an unnamed allowlist. So this additional check makes sense.

Could you add error handling like the other `session.query` calls? There's some code above this with examples.

**Comment by galmasi** [comment]
I think the feeling among the reviewers was to minimize the disruption with this PR -- it's already had too long a life -- and postpone complex fixes. 

Without troubling this PR any further, @maugustosilva @mdrocco and myself are finding issues related to operation of IMA at scale and against databases != sqlite. We don't want to hurry and make things worse. We are collecting data for opening a larger issue that can serve as a basis for fixes.

File: keylime/cloud_verifier_tornado.py
Comment (stefanberger) [inline keylime/cloud_verifier_tornado.py:572]: you may want to squash this commit into the previous one ...

**Comment by stefanberger** [comment]
LGTM

**Comment by stefanberger** [comment]
To get rid of the test failure you will probably need to rebase on master.

**Comment by aplanas** [comment]
Can you drop the merge commit and make a rebase? I think the history would be a bit more clear : ))

Otherwise LGTM

**Comment by galmasi** [comment]
@aplanas  Dropped the merge and rebased instead.

**Review by aplanas — APPROVED** [review]

**Review by THS-on — APPROVED** [review]

**Review by mpeters — APPROVED** [review]

**Pull request merged**
